### ISSUE TITLE
prebuilt binary "panick" crash

### ISSUE THREAD

**Issue by marlop352** [code-inflation/cfspeedtest#112]
result with RUST_LOG=debug and RUST_BACKTRACE=full
```
Starting Cloudflare speed test
[2024-07-13T03:49:45Z DEBUG reqwest::connect] starting new connection: https://speed.cloudflare.com/
[2024-07-13T03:49:45Z DEBUG hyper_util::client::legacy::connect::dns] resolving host="speed.cloudflare.com"
[2024-07-13T03:49:45Z DEBUG hyper_util::client::legacy::connect::http] connecting to 104.16.60.8:443
[2024-07-13T03:49:45Z DEBUG hyper_util::client::legacy::connect::http] connected to 104.16.60.8:443
[2024-07-13T03:49:45Z DEBUG rustls::client::hs] No cached session for DnsName("speed.cloudflare.com")
[2024-07-13T03:49:45Z DEBUG rustls::client::hs] Not resuming any session
[2024-07-13T03:49:45Z DEBUG rustls::client::hs] Using ciphersuite TLS13_AES_256_GCM_SHA384
[2024-07-13T03:49:45Z DEBUG rustls::client::tls13] Not resuming
[2024-07-13T03:49:45Z DEBUG rustls::client::tls13] TLS1.3 encrypted extensions: [ServerNameAck, Protocols([ProtocolName(687474702f312e31)])]
[2024-07-13T03:49:45Z DEBUG rustls::client::hs] ALPN protocol is Some(b"http/1.1")
[2024-07-13T03:49:45Z DEBUG hyper_util::client::legacy::pool] pooling idle connection for ("https", speed.cloudflare.com)
thread 'main' panicked at src\speedtest.rs:305:10:
called `Result::unwrap()` on an `Err` value: ToStrError { _priv: () }
stack backtrace:
   0:     0x7ff69e83f93a - <unknown>
   1:     0x7ff69e8597eb - <unknown>
   2:     0x7ff69e83c841 - <unknown>
   3:     0x7ff69e83f726 - <unknown>
   4:     0x7ff69e8418cf - <unknown>
   5:     0x7ff69e841567 - <unknown>
   6:     0x7ff69e841e0d - <unknown>
   7:     0x7ff69e841cc9 - <unknown>
   8:     0x7ff69e83ffc9 - <unknown>
   9:     0x7ff69e841986 - <unknown>
  10:     0x7ff69e87abb7 - <unknown>
  11:     0x7ff69e87b0b3 - <unknown>
  12:     0x7ff69e55dbeb - <unknown>
  13:     0x7ff69e55d5d7 - <unknown>
  14:     0x7ff69e55a8ef - <unknown>
  15:     0x7ff69e5438c7 - <unknown>
  16:     0x7ff69e541886 - <unknown>
  17:     0x7ff69e54189c - <unknown>
  18:     0x7ff69e8374a2 - <unknown>
  19:     0x7ff69e543abc - <unknown>
  20:     0x7ff69e86c2bc - <unknown>
  21:     0x7ffc5c41257d - BaseThreadInitThunk
  22:     0x7ffc5da2af28 - RtlUserThreadStart
```
Windows 11 23H2 - 22631.3880

--------------------------------------------------------------------------------------------------

also on linux with docker (no special env vars)
```
Starting Cloudflare speed test
thread 'main' panicked at src/speedtest.rs:305:10:
called `Result::unwrap()` on an `Err` value: ToStrError { _priv: () }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
(Proxmox) Linux box 6.8.8-2-pve #1 SMP PREEMPT_DYNAMIC PMX 6.8.8-2 (2024-06-24T09:00Z) x86_64 x86_64 x86_64 GNU/Linux
Docker version 27.0.3, build 7d4bcd8

**Comment by code-inflation** [comment]
I can't reproduce the issue on Linux (native or in docker) with `cfspeedtest` version `v1.2.4`. But according to the backtrace it seems that the issue is converting the header value to a string.

**Comment by marlop352** [comment]
whenever there is a build with this fix for windows I can test and report the results

**Comment by code-inflation** [comment]
I have a fix ready to be tested from this MR: https://github.com/code-inflation/cfspeedtest/pull/113
Please checkout the branch `bugfix/fix-extrac-header-value` and run `cargo build --release` or use this cross compiled build to test it: [cfspeedtest.zip](https://github.com/user-attachments/files/16239282/cfspeedtest.zip)

**Comment by marlop352** [comment]
Using your cross compiled build, It ran but the city field seems to not be populated properly.
I also got a crash when my network connection borked itself around the same time I started a test but I forgot to save the log from it 🤦🏻‍♂️

Here are some results with some redactions (IP, ASN and COLO)

Clean run:
```
Starting Cloudflare speed test
City: City N/A
Country: BR
Ip: XXX.XXX.XXX.XXX
Asn: XXXXX
Colo: XXX
latency test    [==============================]
Avg GET request latency 49.61 ms (RTT excluding server processing time)

Download 100KB  [==============================]   27.10 mbit/s | 100KB in   29ms -> status: 200 OK
Download 1MB    [==============================]  230.23 mbit/s |   1MB in   34ms -> status: 200 OK
Download 10MB   [==============================]  224.64 mbit/s |  10MB in  356ms -> status: 200 OK
Download 25MB   [==============================]  214.66 mbit/s |  25MB in  931ms -> status: 200 OK
Upload 100KB    [==============================]    9.94 mbit/s | 100KB in   80ms -> status: 200 OK
Upload 1MB      [==============================]  122.89 mbit/s |   1MB in   65ms -> status: 200 OK
Upload 10MB     [==============================]  215.85 mbit/s |  10MB in  370ms -> status: 200 OK
Upload 25MB     [==============================]  209.28 mbit/s |  25MB in  955ms -> status: 200 OK

Summary Statistics
Type     Payload |  min/max/avg in mbit/s
Download  100KB  |  min 27.10   max 85.69   avg 64.97
Download  1MB    |  min 93.03   max 262.14  avg 220.51
Download  10MB   |  min 160.93  max 265.09  avg 213.73
Download  25MB   |  min 174.49  max 247.05  avg 219.99
Upload    100KB  |  min 6.93    max 14.10   avg 12.37
Upload    1MB    |  min 50.02   max 122.89  avg 103.91
Upload    10MB   |  min 206.72  max 218.77  avg 212.88
Upload    25MB   |  min 145.71  max 226.64  avg 199.05
```

run with RUST_LOG=debug and RUST_BACKTRACE=full:
```
Starting Cloudflare speed test
[2024-07-15T21:38:51Z DEBUG reqwest::connect] starting new connection: https://speed.cloudflare.com/
City: City N/A
Country: BR
Ip: XXX.XXX.XXX.XXX
Asn: XXXXX
Colo: XXX
latency test    [==============================]
Avg GET request latency 31.22 ms (RTT excluding server processing time)

[2024-07-15T21:38:53Z DEBUG cfspeedtest::speedtest] getting payload iterations for max_payload_size M25
[2024-07-15T21:38:53Z DEBUG cfspeedtest::speedtest] running tests for payload_size 100000
Download 100KB  [==============================]  325.72 mbit/s | 100KB in    2ms -> status: 200 OK
[2024-07-15T21:38:54Z DEBUG cfspeedtest::speedtest] running tests for payload_size 1000000
Download 1MB    [==============================]  193.51 mbit/s |   1MB in   41ms -> status: 200 OK
[2024-07-15T21:38:55Z DEBUG cfspeedtest::speedtest] running tests for payload_size 10000000
Download 10MB   [==============================]  255.62 mbit/s |  10MB in  312ms -> status: 200 OK
[2024-07-15T21:39:00Z DEBUG cfspeedtest::speedtest] running tests for payload_size 25000000
Download 25MB   [==============================]  188.92 mbit/s |  25MB in 1058ms -> status: 200 OK
[2024-07-15T21:39:09Z INFO  cfspeedtest::speedtest] Exceeded threshold
[2024-07-15T21:39:09Z DEBUG cfspeedtest::speedtest] running tests for payload_size 100000
Upload 100KB    [==============================]   13.58 mbit/s | 100KB in   58ms -> status: 200 OK
[2024-07-15T21:39:10Z DEBUG cfspeedtest::speedtest] running tests for payload_size 1000000
Upload 1MB      [==============================]   41.70 mbit/s |   1MB in  191ms -> status: 200 OK
[2024-07-15T21:39:12Z DEBUG cfspeedtest::speedtest] running tests for payload_size 10000000
Upload 10MB     [==============================]   41.66 mbit/s |  10MB in 1920ms -> status: 200 OK
[2024-07-15T21:39:32Z INFO  cfspeedtest::speedtest] Exceeded threshold

Summary Statistics
Type     Payload |  min/max/avg in mbit/s
[2024-07-15T21:39:32Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [186.35855385762207, 175.37706068046302, 281.4820027444495, 228.07617744326606, 288.70443883074705, 291.6833776935137, 288.8608051994945, 249.15908807773764, 316.40563202025, 325.71963682260497]
Download  100KB  |  min 175.38  max 325.72  avg 263.18
[2024-07-15T21:39:32Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [102.14621975993084, 259.2865083506461, 225.370523225841, 164.53491122313196, 225.06175131801785, 192.8245176374176, 283.3031733496705, 161.03869961250064, 140.30681592973434, 193.51108939486664]
Download  1MB    |  min 102.15  max 283.30  avg 194.74
[2024-07-15T21:39:32Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [170.31999507775214, 262.2174484078976, 257.26467196503256, 200.4001489975108, 245.27966941206157, 177.92917573642106, 187.80197971456917, 214.7094779473237, 234.28785132561532, 255.61628927584223]
Download  10MB   |  min 170.32  max 262.22  avg 220.58
[2024-07-15T21:39:32Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [246.12770360514324, 263.77620699046514, 218.48143170008265, 232.8688841062795, 244.6925572598936, 240.9259458692403, 246.05348661481335, 242.29989118311886, 234.0925618375985, 188.91538951991876]
Download  25MB   |  min 188.92  max 263.78  avg 235.82
[2024-07-15T21:39:32Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [6.987534238917771, 11.73583228322084, 7.086977590090989, 11.869788421021395, 13.023080153802576, 12.961468793643697, 12.48289453358445, 12.93663152754128, 13.635658610833191, 13.582158476625105]
Upload    100KB  |  min 6.99    max 13.64   avg 11.63
[2024-07-15T21:39:32Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [31.959564758667334, 40.671506914918766, 41.45279614654807, 39.89567281558724, 41.30245217821388, 41.417900402374904, 39.91069980917697, 42.59721616543055, 42.068331591003265, 41.702265631963954]
Upload    1MB    |  min 31.96   max 42.60   avg 40.30
[2024-07-15T21:39:32Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [41.347442872563974, 31.87067136525351, 38.764195508964434, 44.026449990493035, 42.71508392953356, 37.569859392453104, 48.75364951537653, 51.34132418259798, 40.27781015386476, 41.6562330820369]
Upload    10MB   |  min 31.87   max 51.34   avg 41.83
```

**Comment by marlop352** [comment]
Running the traffic trough a tunnel to a datacenter worked too and the city field populated correctly, so you probably can ignore the city thing.

run with RUST_LOG=debug and RUST_BACKTRACE=full (same redactions as my previous runs):
```
Starting Cloudflare speed test
[2024-07-15T21:53:34Z DEBUG reqwest::connect] starting new connection: https://speed.cloudflare.com/
City: Vinhedo
Country: BR
Ip: XXX.XXX.XXX.XXX
Asn: XXXXX
Colo: XXX
latency test    [==============================]
Avg GET request latency 33.90 ms (RTT excluding server processing time)

[2024-07-15T21:53:37Z DEBUG cfspeedtest::speedtest] getting payload iterations for max_payload_size M25
[2024-07-15T21:53:37Z DEBUG cfspeedtest::speedtest] running tests for payload_size 100000
Download 100KB  [==============================]  278.76 mbit/s | 100KB in    2ms -> status: 200 OK
[2024-07-15T21:53:38Z DEBUG cfspeedtest::speedtest] running tests for payload_size 1000000
Download 1MB    [==============================]  141.73 mbit/s |   1MB in   56ms -> status: 200 OK
[2024-07-15T21:53:40Z DEBUG cfspeedtest::speedtest] running tests for payload_size 10000000
Download 10MB   [==============================]  180.40 mbit/s |  10MB in  443ms -> status: 200 OK
[2024-07-15T21:53:45Z INFO  cfspeedtest::speedtest] Exceeded threshold
[2024-07-15T21:53:45Z DEBUG cfspeedtest::speedtest] running tests for payload_size 100000
Upload 100KB    [==============================]   14.53 mbit/s | 100KB in   55ms -> status: 200 OK
[2024-07-15T21:53:46Z DEBUG cfspeedtest::speedtest] running tests for payload_size 1000000
Upload 1MB      [==============================]  106.00 mbit/s |   1MB in   75ms -> status: 200 OK
[2024-07-15T21:53:47Z DEBUG cfspeedtest::speedtest] running tests for payload_size 10000000
Upload 10MB     [==============================]  203.44 mbit/s |  10MB in  393ms -> status: 200 OK
[2024-07-15T21:53:51Z DEBUG cfspeedtest::speedtest] running tests for payload_size 25000000
Upload 25MB     [==============================]  165.21 mbit/s |  25MB in 1210ms -> status: 200 OK
[2024-07-15T21:54:05Z INFO  cfspeedtest::speedtest] Exceeded threshold

Summary Statistics
Type     Payload |  min/max/avg in mbit/s
[2024-07-15T21:54:05Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [220.22187353758915, 173.97733945153647, 200.68735418809425, 235.16270319527325, 216.39166892074655, 138.85272932396077, 210.64827005108222, 243.34600760456274, 228.92462656670293, 278.7553573295237]
Download  100KB  |  min 138.85  max 278.76  avg 214.70
[2024-07-15T21:54:05Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [133.4227265601286, 144.20546394502887, 105.318588730911, 97.31436676325117, 85.82567517985844, 172.15333697724563, 108.2516146404896, 107.96396162960802, 164.41757130584546, 141.73063738039264]
Download  1MB    |  min 85.83   max 172.15  avg 126.06
[2024-07-15T21:54:05Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [139.49357902337405, 125.95822721352691, 196.36436291177915, 165.6210115096252, 196.68729424665077, 178.93753156569892, 161.7775469751441, 199.3846988194432, 177.2260141204827, 180.39889352338656]
Download  10MB   |  min 125.96  max 199.38  avg 172.18
[2024-07-15T21:54:05Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [8.108795712068828, 14.38135026497638, 15.039488055826581, 15.246168447292092, 15.07144808357118, 14.116866479147625, 14.31086051204259, 13.549883047571946, 14.394029356622873, 14.531002802667167]
Upload    100KB  |  min 8.11    max 15.25   avg 13.87
[2024-07-15T21:54:05Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [58.86512454756633, 100.59590499219753, 85.99515202330468, 93.74487332723992, 65.26728175640783, 96.17534689245431, 99.18187352079534, 86.08333727881963, 101.61852912260022, 105.99831197688178]
Upload    1MB    |  min 58.87   max 106.00  avg 89.35
[2024-07-15T21:54:05Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [208.17530442385532, 190.5098018483737, 197.68032028165493, 198.78527289366505, 201.30668167072466, 203.56477545660215, 203.1163120171227, 206.16744765472933, 209.30879955124192, 203.442346219202]
Upload    10MB   |  min 190.51  max 209.31  avg 202.21
[2024-07-15T21:54:05Z DEBUG cfspeedtest::measurements] calc_stats for mbit_measurements [210.18402557267004, 90.67429394647532, 111.22758493046413, 128.89248040624847, 133.9509005652996, 149.12539447394974, 142.20669957092684, 156.06190882285858, 180.17016531603605, 165.20617524162435]
Upload    25MB   |  min 90.67   max 210.18  avg 146.77
```

**Comment by code-inflation** [comment]
Thanks for trying it out. Currently we just use the `cf-meta-city` header field to populate the city name. So my suspicion would be that CF doesn't populate the filed from their end when you are using the connection that you tested with here: https://github.com/code-inflation/cfspeedtest/issues/112#issuecomment-2229499471

You could print out the headers and look for the `cf-meta-city` field and check if it is populated:
```sh
curl -v https://speed.cloudflare.com/__down
```

**Comment by code-inflation** [comment]
Fix is released in `1.2.5` https://github.com/code-inflation/cfspeedtest/releases/tag/v1.2.5

**Comment by marlop352** [comment]
Thanks for the fix and updated release.

cf-meta-city is populated, but it has a special char in it that shows as "Ã¢" on the output of that curl command. the special char is "â".
While it's interesting having the city name I don't particularly care about it.
But I can help with any testing/debugging needed if you care and I don't need to compile it 😅
